accepted/0000-caasp-integration.md

```diff
@@ -0,0 +1,422 @@
+
+Example:
+
+```
+bootstrap_first_control_plane:
+  cmd.run:
+    - name: skuba node bootstrap --user sles --sudo --target <node IP> <node name>
+
+bootstrap_additional_control_plane:
+  cmd.run:
+    - name: skuba node join --role master --user sles --sudo --target <node IP> <node name>
+
+bootstrap_worker:
+  cmd.run:
+    - name: skuba node join --role worker --user sles --sudo --target <node IP> <node name>
+```
+
+Uyuni/SUSE Manager will check the return code of the `skuba` command and, in case of errors, show the raw output.
+
+### Manage a CaaS Platform cluster using Uyuni/SUSE Manager
+
+#### Show node status
+
+Individual nodes that constitute the cluster can be accessed and managed as already happen with Uyuni/SUSE Manager, with the restrictions described above.
+Hardware details of the node will be shown.
+
+From a Uyuni/SUSE Manager point of view, it would make sense to create a system group to identify the nodes part of the cluster.
+
+#### Show cluster status
+
+Uyuni/SUSE Manager shows the status of the cluster by invoking `skuba cluster status` on the Management Node (via Salt).
+
+E.g. The output is presented in the "CaaS Platform > cluster name" section:
+
+```
+NAME         OS-IMAGE                              KERNEL-VERSION        CONTAINER-RUNTIME   HAS-UPDATES   HAS-DISRUPTIVE-UPDATES
+cp0       SUSE Linux Enterprise Server 15 SP1   4.12.14-110-default   cri-o://1.13.3      <none> <none>
+worker0   SUSE Linux Enterprise Server 15 SP1   4.12.14-110-default   cri-o://1.13.3      <none> <none>
+```
+
+#### Removing nodes from the cluster
+
+A node can also be removed from the cluster:
+
+- temporarily (drained and cordoned off): [using `kubectl`](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#use-kubectl-drain-to-remove-a-node-from-service) from the Management Node via Salt.
+- permanently: by invoking `skuba node remove <nodename>` via Salt. NOTE: this removal is permanent and the cluster cannot join any cluster in the future. A complete reinstallation is required and Uyuni/SUSE Manager must offer the user to re-provision the node from scratch, e.g. via AutoYaST.
+
+Example state:
+
+```
+remove_node:
+      cmd.run:
+        - name: skuba node remove <nodename>
+```
+
+NOTE: Other `skuba` options can be implemented in the future (e.g. resetting and reconfiguring a node).
+
+#### Labeling nodes
+
+Uyuni/SUSE Manager can apply node labels to describe what is a node used for. This possiblity is offered to the user and it translated to a `kubectl label node nodename description` issued by the Management Node.
+
+#### Cluster upgrade
+
+All the cluster updates must be handled by `skuba`. NOTE: it is recommended to upgrade the whole cluster machine by machine, all of them at the same time.
```

Comment:
Sorry if I'm misunderstanding/missing  anything in here. But is it possible to have different options for doing cluster upgrade, either:

1. one node by one node, one at a time while the rest cluster nodes and cluster services on them remain running, which is so-called "Cluster Rolling Upgrade" for HA:

https://www.suse.com/documentation/sle-ha-15/book_sleha_guide/data/sec_ha_migration_terminology.html

or:
2. Shut down cluster services on all cluster nodes and upgrade all of them in parallel at the same time, which is also so-called "Cluster Offline Upgrade" for HA

?

HA has to support both ways. 

And either way, is it possible to invoke pre and post jobs before/after upgrade of the whole cluster, and even between the upgrades of cluster nodes? For example operations on cluster services/resources on cluster nodes?

Thanks.

Reply:
The section of the RFC you are referring to is related to CaaSP but in the link mentions SLE-HA, I am bit confused. Would you care to clarify please?

Reply:
@mbologna  @paususe asked us to review this as a Generic Cluster integration for SUMA, so the questions raised by @gao-yan refers to HA cluster upgrade options we have. If this is only focused for CaaSP, then maybe we need to review this part on a different RFC.

Reply:
@diegoakechi : yes you are very welcome to do and I am eager to know your inputs. The integration part for a generic cluster is in the first section of this RFC (at the top).
Should we clarify the issue that @gao-yan raised in the generic section of the RFC?

Reply:
@diegoakechi it's this RFC just a different section.
@mbologna as this is no longer just a caasp4-integration rfc but a clustwr awareness and caasp4 integrstion rrc, please make the necessary changes according to @gao-yan 's comments but in the right section. No need to get to the gory details at this point but at least, this rfc should not contradict the HA casa, that suse manager will have to support in the future. 

IMHO this "update policy" cluster attribute makes total sense for all kinds of clusters (caasp, virtual machines, ha, etc). No need to define all the possible values atm, though, just the one(s) needed for caasp4. HA's and openstack and other cluster update policies we can discuss in a future HA cluster support rfc.

The pre and post actions also seem sensible. Does caasp need any specifics implemented?

Reply:
> 1)one node by one node, one at a time while the rest cluster nodes and cluster services on them remain running, which is so-called "Cluster Rolling Upgrade" for HA:

Sure, that's how every cluster will change/update itself. But all of this **must be orchestrated** by the cluster (admin node) and **not** from SUSE Manager.

Reply:
> 2)shut down cluster services on all cluster nodes and upgrade all of them in parallel at the same time, which is also so-called "Cluster Offline Upgrade" for HA

The whole purpose of having a cluster is that you **never** shut it down for any kind of config changes.

Reply:
> > 2)shut down cluster services on all cluster nodes and upgrade all of them in parallel at the same time, which is also so-called "Cluster Offline Upgrade" for HA
> 
> The whole purpose of having a cluster is that you **never** shut it down for any kind of config changes.
Not 100% true. It comes to a point that base system and HA stack need to do for example a major version upgrade, for example from SLE 15 -> SLE 16,  which might intentionally introduce changes that are not backward-compatible so that it's no longer possible to run the cluster with the nodes having these mixed versions during upgrade:
https://www.suse.com/documentation/sle-ha-15/book_sleha_guide/data/sec_ha_migration_upgrade.html#sec_ha_migration_upgrade_offline

But anyway probably it wouldn't necessarily introduce much more effort to make shutdown of whole cluster possible.